### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'update'

### ISSUE THREAD

**Issue by HerbertArthur** [awslabs/autogluon#488]
I am trying to run this code like below:

```python
#-*- coding:utf-8 -*-

import autogluon as ag
from autogluon import ImageClassification as task

#filename=ag.download("https://autogluon.s3.amazonaws.com/datasets/shopee-iet.zip")
#ag.unzip(filename)
dataset = task.Dataset('data/train')
test_dataset = task.Dataset('data/test', train=False)
classifier = task.fit(dataset, epochs=3, ngpus_per_trial=1, verbose=True)
print(">>>>")
HBox(children=(FloatProgress(value=0.0, max=2.0), HTML(value="")))
print("Top 1 val accuracy: %.3f" % classifier.results["best_reward"])
```

Then I got error, the output is following:

>(mxnet) [root@46acffeb04ca autogluon]# python classfication_opt.py 
>scheduler: FIFOScheduler(
>DistributedResourceManager{
>(Remote: Remote REMOTE_ID: 0, 
><Remote: 'inproc://172.17.0.15/14336/1' processes=1 threads=28, memory=202.71 GB>, Resource: NodeResourceManager(28 CPUs, 4 GPUs))
>})
>
>Starting Experiments
>Num of Finished Tasks is 0
>Num of Pending Tasks is 2
>[Epoch 3] Validation: 0.388: 100%|████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:27<00:00,  9.06s/it]
>100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:31<00:00, 15.95s/it]
>[Epoch 3] Validation: 0.425: 100%|████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:27<00:00,  9.03s/it]
>[Epoch 3] training: accuracy=0.441: 100%|█████████████████████████████████████████████████████████████████████████████| 3/3 [00:27<00:00,  9.30s/it]
>Traceback (most recent call last):
>  File "classfication_opt.py", line 10, in <module>
>    classifier = task.fit(dataset, epochs=3, ngpus_per_trial=1, verbose=False)
>  File "/root/anaconda3/envs/mxnet/lib/python3.7/site-packages/autogluon/task/image_classification/image_classification.py", line 294, in fit
>    results = BaseTask.run_fit(train_image_classification, search_strategy, scheduler_options)
>  File "/root/anaconda3/envs/mxnet/lib/python3.7/site-packages/autogluon/task/base/base_task.py", line 63, in run_fit
>    results.update(best_reward=best_reward,
>AttributeError: 'NoneType' object has no attribute 'update'

My envs are: `mxnet-cu100 1.5.1.post0, autogluon 0.0.9`

**Comment by zhreshold** [comment]
@HerbertArthur Cannot reproduce, can you try upgrade autogluon?

**Comment by yongzpp** [comment]
@HerbertArthur did you manage to resolve? I encountered the same issue

**Comment by mseeger** [comment]
Does this still happen with the latest version?

**Comment by yongzpp** [comment]
I am using 0.0.10 version

**Comment by mseeger** [comment]
@zhreshold 

What happens here:
In BaseTask.run_fit, there is a call to scheduler.run_with_config(best_config). This returns None. It itself calls TasKScheduler.run_job, which returns job.result(), so this must be None.

I'd check whether the best_config passed to run_with_config is really valid. But I also do not see an obvious reason why any of this happens.

**Comment by zhreshold** [comment]
tracking the result down into the dask, I found that it's directly returned by `dask.Future.result()` which is supposed to be the return value of image classification training pipeline: https://github.com/awslabs/autogluon/blob/master/autogluon/task/image_classification/pipeline.py#L121

So the root cause might be related to the execution of dask.

**Comment by mseeger** [comment]
So we should for sure test for None in BaseTask.run_fit, after scheduler.run_with_config(best_config) has been called (did you check that best_config is really valid?), and fail there with a message.

Even better if we could avoid this happening altogether.

**Comment by FlorianBorn** [comment]
Hi,
I just encountered this error.
It occurs if I use 60 or less Images per class (cats & dogs) and disappers if I use 70 Images per class.
Is it possible, that this error is related to the size of the dataset?

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-7331ea28d8b1> in <module>
      1 # fit
      2 time_limits = 10 * 60 # 10mins
----> 3 classifier = task.fit(dataset, epochs=1, num_trials=1, nthreads_per_trial=ag.get_cpu_count(), verbose=True)

~/anaconda3/envs/autogluon/lib/python3.7/site-packages/autogluon/task/image_classification/image_classification.py in fit(dataset, net, optimizer, loss, split_ratio, batch_size, input_size, epochs, final_fit_epochs, ensemble, metric, nthreads_per_trial, ngpus_per_trial, hybridize, scheduler_options, search_strategy, search_options, plot_results, verbose, num_trials, time_limits, resume, output_directory, visualizer, dist_ip_addrs, auto_search, lr_config, tricks, **kwargs)
    311         results = BaseTask.run_fit(
    312             train_image_classification, search_strategy, scheduler_options,
--> 313             plot_results=plot_results)
    314         args = sample_config(train_image_classification.args, results['best_config'])
    315 

~/anaconda3/envs/autogluon/lib/python3.7/site-packages/autogluon/task/base/base_task.py in run_fit(cls, train_fn, search_strategy, scheduler_options, plot_results)
     71             scheduler.get_training_curves(filename=plot_training_curves, plot=True, use_legend=False)
     72         record_args = copy.deepcopy(args)
---> 73         results.update(best_reward=best_reward,
     74                        best_config=best_config,
     75                        total_time=total_time,

AttributeError: 'NoneType' object has no attribute 'update'
```

**Comment by Innixma** [comment]
@zhreshold What happens if training fails for the model? Does it return None? If so, we should raise a proper error stating the model exception.

**Comment by jamiekang** [comment]
Hi,
I am also having this error.
With the same script, there was no error before the recent update of Autogluon (last Friday).
How can we avoid this error?

```
Saving Training Curve in checkpoint/plot_training_curves.png
Traceback (most recent call last):
  File "ag.py", line 37, in <module>
    num_trials = 10
  File "/home/ubuntu/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/autogluon/task/image_classification/image_classification.py", line 313, in fit
    plot_results=plot_results)
  File "/home/ubuntu/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/autogluon/task/base/base_task.py", line 73, in run_fit
    results.update(best_reward=best_reward,
AttributeError: 'NoneType' object has no attribute 'update'
```
This is my task.fit().
```
classifier = task.fit(dataset,
                      time_limits=time_limits,
                      ngpus_per_trial = num_gpu,
                      verbosity = 4,
                      epochs = 50,
                      batch_size = 128,
                      verbose = True,
                      plot_results = True,
                      auto_search = True,
                      num_trials = 10
                     )
```

**Comment by FlorianBorn** [comment]
@jamiekang 
What's the size of your dataset (images per class)? Maybe try to set your batch_size to be smaller then your dataset. That worked for me.

**Comment by jamiekang** [comment]
> @jamiekang
> What's the size of your dataset (images per class)? Maybe try to set your batch_size to be smaller then your dataset*. That worked for me.
> 
> Edit:
> *smaller then the size of your class with the least samples

The batch size is 128, and the smallest class size is about 2000.

**Comment by Innixma** [comment]
@jamiekang Can you try installing mxnet 1.7 based on the instructions in https://autogluon.mxnet.io/install.html#installation-faq ? Because mxnet 1.7 has not officially released quite yet, simply doing pip install --upgrade mxnet won't quite do the trick.

**Comment by jamiekang** [comment]
> @jamiekang Can you try installing mxnet 1.7 based on the instructions in https://autogluon.mxnet.io/install.html#installation-faq ? Because mxnet 1.7 has not officially released quite yet, simply doing pip install --upgrade mxnet won't quite do the trick.

I followed that instruction. And I tried both cuda 10.0 and 10.1. The same error.

**Comment by zhreshold** [comment]
did you encounter have error message or traceback before the error was actually raised?

**Comment by jamiekang** [comment]
> did you encounter have error message or traceback before the error was actually raised?

I cannot see anything relevant. You can see the whole log file here -> [link](https://jk-s3-iad.s3.amazonaws.com/temp/ag.out.gz)

**Comment by liangfu** [comment]
https://github.com/awslabs/autogluon/blob/132e61736d392fe50650d148b8d01d227117e17b/core/src/autogluon/core/task/base/base_task.py#L67-L73

Since L67 here returns None in results, L73 couldn't get the update function from results dict. I think this might be the root cause of the issue.

**Comment by mseeger** [comment]
Yeah, but why does it return None? One has to go a little deeper, as this is for the most time not happening

**Comment by MatheCore** [comment]
Do anyone know how to fix this error? I have the same error in almost the same code. I am working with autogluon and running my code on google colab. But when I don't use the GPU and run my code with FashionMNIST dataset, i get no error. But i have to use the GPU becuase my dataset is very big and i also need to test no FashionMNIST dataset.
thanks for help.

**Comment by liangfu** [comment]
Previously, I had this error when I was trying to perform image classification training with image size less than 224x224 or the number of images are too small for training (less than 100 samples). I increased the number of samples, and resize images to 224x224, and the error disappeared. Still not quite sure about what is the root cause under the hood.

**Comment by Innixma** [comment]
@MatheCore Can you try to use the pre-release version of AutoGluon? The fix in #744 is not available in the stable release. Example colab notebook of pre-release can be found [here](https://github.com/awslabs/autogluon/issues/818#issuecomment-744710997).

Pre-release docs are available at https://auto.gluon.ai/dev/index.html

**Comment by MatheCore** [comment]
> @MatheCore Can you try to use the pre-release version of AutoGluon? The fix in #744 is not available in the stable release. Example colab notebook of pre-release can be found [here](https://github.com/awslabs/autogluon/issues/818#issuecomment-744710997).
> 
> Pre-release docs are available at https://auto.gluon.ai/dev/index.html

At that time i tried very different things but nothing worked. Then i worked with another type of data for my research. But for now i really need to work with images. I have this code now:
`import autogluon as ag`

`from autogluon import ImageClassification as task`

`#filename = ag.download('/content/cat-and-dog.zip')`

`#ag.unzip(filename)`

`dataset = task.Dataset('training_set/training_set')`

`print(dataset)`

`time_limits = 10 * 60 # 10mins`

`classifier = task.fit(dataset, time_limits=time_limits,ngpus_per_trial=1)`


and i get the following error: 

> RecursionError: maximum recursion depth exceeded while calling a Python object

> During handling of the above exception, another exception occurred:

> Traceback (most recent call last):

>   File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap

>     self.run()

 

>  File "/usr/local/lib/python3.6/dist-packages/autogluon/utils/custom_process.py", line 20, in run

  

>   self._cconn.send((e, tb))

>   File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send

>     self._send_bytes(_ForkingPickler.dumps(obj))

>   File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps

>     cls(buf, protocol).dump(obj)

> RecursionError: maximum recursion depth exceeded while pickling an object


Training_set/Trainig_set is basically the folder that includes two folders in it, one for cats images and one for dogs images. 
when i try to print the dataset, it just print "AutoGluonObject"
I am actually still new in this. so maybe i am just doing biginners mistake. 
I tried most of the ways i find on google but nothings work. I really appreciate any help. Thank you!

**Comment by Innixma** [comment]
I am guessing this has been fixed in pre-release and for the upcoming v0.1 release due to code overhauls. If this issue still exists in pre-release / v0.1+, please open a new issue.
